content/ADR-146-unity-2021-material-upgrade.md

```diff
@@ -0,0 +1,77 @@
+
+
+The full details of the algorithm can be found [here](https://github.com/decentraland/unity-renderer/blob/dev/unity-renderer/Assets/Scripts/MainScripts/DCL/Helpers/MaterialHelpers/SRPBatchingHelper/SRPBatchingHelper.cs) but the conclusion is that we are gonna have this range of render queues depending on the render target:
+
+```
+Opaque 2000 + crc[0, Permutations of shaderName + keywords]  + queueOffset (150, 300 or 450)
+
+Alpha Test 2600 + crc[0, Permutations of shaderName + keywords]  + queueOffset (150, 300 or 450)
+
+Transparent ALWAYS 3000
+```
+
+That algorithm shows some overlapping in some extreme cases.
+- Opaque materials will always be in a render queue < 2600 if we have less than 150 shader variants permutations. A quick test showed that walking around different scenes for an hour accumulated a total of 15 permutations, so assuming `renderQueue < 2600 => Opaque` is safe.
+- Transparent materials will always be in `render queue == 3000` by the algorithm (we dont do further optimization because we need clip space z sorting to render them properly).
+- Alpha Tested wil always be in a `render queue >= 2600`. The only conflict is in the render queue 3000 (as we saw before, it's reserved to transparent materials). This case can occur if:
+  - An alpha tested material with queue offset of 150 (`Cull Off`) is optimized at exactly 250 permutations (`2600 + 150 + 250 = 3000`). It's highly unlikely to have so many permutations, we can ignore this.
+  - An alpha tested material with queue offset of 300 (`Cull Front`) is optimized at exactly 100 permutations (`2600 + 300 + 100 = 3000`). That also goes way above our measurements in permutations, and in any case `Cull Front` is the least used cull setting.
+
+So we can effectively use the render queue to spot the render target of the material!
+
+## Conclusion
+For the serialized properties, we can retrieve anything that was there before due to the nature of serialization.
+For the render target, we can use the render queue that the SRP Batcher defines to know which kind of material are we dealing with. This will allow us to set the keywords needed for the new URP material.
+
+## Future work
+Unity can change the Asset Bundle serialization in any new version and in that case, no manual conversion will make it compatible. In the same time we shouldnt sacrifice potential improvements in performance, new language features or new systems due to being stuck in an older AB serialization.
+
+At some point we will need to reconvert the world Asset Bundles and by then we will need versioning, tools to make the process painless and a smart way of storing the assets in a way that we have the deprecated version as less time as possible.
+
+Currently we just upgraded to Unity 2021 and it will accompany us for a while, the above scenario is far way from now but nonetheless we should start preparing ourselves for it.
```

Comment:
```suggestion
Currently, we just upgraded to Unity 2021 and it will accompany us for a while, the above scenario is far away from now. Nonetheless, we should start preparing ourselves for it.
```